feat: vercel edge middleware support (#7532)
withastro/astro: packages/integrations/vercel/README.md

});
```

### Vercel Edge Middleware

> You can use Vercel edge middleware to intercept a request and redirect before sending a response. Vercel middleware can run for Edge, SSR, and Static deployments. You don't need to install `@vercel/edge` to write middleware, but you do need to install it to use features such as geolocation. For more information see [Vercel’s middleware documentation](https://vercel.com/docs/concepts/functions/edge-middleware).

Reviewer: ```suggestion
You can use Vercel Edge middleware to intercept a request and redirect before sending a response. Vercel middleware can run for Edge, SSR, and Static deployments.  You may not need to install this package for your middleware. `@vercel/edge` is only required to use some middleware features such as geolocation. For more information see [Vercel’s middleware documentation](https://vercel.com/docs/concepts/functions/edge-middleware).
```
Maybe something more of this style?

> **Warning**
> **Trying to rewrite?** Currently rewriting a request with middleware only works for static files.

### Vercel Edge Middleware with Astro middleware

The `@astrojs/vercel/serverless` adapter can automatically create the vercel edge middleware when you have an Astro middleware in you code base.

> This is an opt-in feature, and the `build.excludeMiddleware` option needs to be set to `true`:

Reviewer 2: Discussion: I talked with Matthew a bit on this. I'm still not sure configuring this option to control an integration behaviour makes sense from an API standpoint.

I think it's more natural if they can configure as:

```js
vercel({
  edgeMiddleware: true
})
```

because `build.excludeMiddleware` can't really be used for all integrations. Maybe we can polish this up in Astro 3.0 again, where I think it's nicer if this can only be configured from integration hooks.

Author: So you think that when passing `edgeMiddleware` to the integration, the integration itself should update the configuration and enable `build.split`?

Reviewer 2: Yeah, but I'm not sure if it's too late now as the same goes for `build.split`. Maybe better to keep this as-is and change it again in the next major.

In the future, we can move the new APIs to the `astro:config:setup` perhaps, similar to `addClientDirectives()`. Maybe named `setBuildMode({ excludeMiddleware: boolean })`.

I think `build.split` could be something always enabled in general though.

```js
// astro.config.mjs
import {defineConfig} from "astro/config";
import vercel from "@astrojs/vercel";
export default definConfig({
   output: "server",
   adapter: vercel(),
   build: {
       excludeMiddleware: true
   }
})
```

Optionally, you can create a known file called `vercel-edge-middleware.(js|ts)` in the [`srcDir`](https://docs.astro.build/en/reference/configuration-reference/#srcdir) folder to create [`Astro.locals`](https://docs.astro.build/en/reference/api-reference/#astrolocals).

```js
// src/vercel-edge-middleware.js

export default function({ request, context }) {
    // do something with request and context
    return {
        title: "Spider-man's blog"
    }
}

> ```

Reviewer 2: Question: Is there a way to _type_ this function? e.g.
```ts
import type { EdgeMiddlewareHandler } from '@astrojs/vercel/serverless`
```

Author: Yes, there is, I will update the example :)

> **Warning**
> **Trying to rewrite?** Currently rewriting a request with middleware only works for static files.

### Vercel Edge Middleware with Astro middleware

> The `@astrojs/vercel/serverless` adapter can automatically create the vercel edge middleware when you have an Astro middleware in you code base.

Reviewer: ```suggestion
The `@astrojs/vercel/serverless` adapter can automatically create the Vercel Edge middleware from an Astro middleware in your code base.
```
Would it be correct to say it like this? That you write it in Astro, but it gets "turned into" Vercel middleware code, executed by them? If so, I think this explains that.

Author: I believe so!

This is an opt-in feature, and the `build.excludeMiddleware` option needs to be set to `true`:

```js
// astro.config.mjs
import {defineConfig} from "astro/config";
import vercel from "@astrojs/vercel";
export default defineConfig({
   output: "server",
   adapter: vercel(),
   build: {
       excludeMiddleware: true
   }
})
```

> Optionally, you can create a known file called `vercel-edge-middleware.(js|ts)` in the [`srcDir`](https://docs.astro.build/en/reference/configuration-reference/#srcdir) folder to create [`Astro.locals`](https://docs.astro.build/en/reference/api-reference/#astrolocals).

Reviewer: ```suggestion
Optionally, you can create a file recognized by Vercel named `vercel-edge-middleware.(js|ts)` in the [`srcDir`](https://docs.astro.build/en/reference/configuration-reference/#srcdir) folder to create [`Astro.locals`](https://docs.astro.build/en/reference/api-reference/#astrolocals).
```
suggest replacing "known" with "recognized by..." whatever understands/recognizes this file!

Author: It's not recognized by Vercel, but it's recognized by the adapter (us) :)

Typings requires the [`@vercel/edge`](https://www.npmjs.com/package/@vercel/edge) package.

```js
// src/vercel-edge-middleware.js
/**
 * 
 * @param options.request {Request}
 * @param options.context {import("@vercel/edge").RequestContext}
 * @returns {object}
 */
export default function({ request, context }) {
    // do something with request and context
    return {

>         title: "Spider-man's blog"

Reviewer: I can't believe you made me spell check this! :joy:

Author: That's what you get from  me when I have to write content that is not "foo", "bar" or "lorem" 🤣

    }
}
```

If you use TypeScript, you can type the function as follows:

```ts
// src/vercel-edge-middleware.ts
import type {RequestContext} from "@vercel/edge";

export default function ({request, context}: { request: Request, context: RequestContext }) {
    // do something with request and context
    return {
        title: "Spider-man's blog"
    }
}
```

The data returned by this function will be passed to Astro middleware.

The function:
- must export a **default** function;
- must **return** an `object`;
- accepts an object with a `request` and `context` as properties;
- `request` is typed as [`Request`](https://developer.mozilla.org/en-US/docs/Web/API/Request);
- `context` is typed as [`RequestContext`](https://vercel.com/docs/concepts/functions/edge-functions/vercel-edge-package#requestcontext);

#### Limitations and constraints 

> When you opt in this feature, there are few constraints to take into consideration:

Reviewer: ```suggestion
When you opt in to this feature, there are few constraints to note:
```

> - the Vercel Edge middleware will always be the **first** function to receive the `Request` and the last function to receive `Response`.
> 
>   This an architectural constraint that follows the [boundaries set by Vercel](https://vercel.com/docs/concepts/functions/edge-middleware).
> - you're allowed to use `request` and `context` only to produce an `Astro.locals` object. 
> 
>   Operations like redirects, etc. should be delegated to Astro middleware
> - `Astro.locals` **must be serializable**, failing to do so will result in a **runtime error**. 
>  
>   This means that you **can't** store complex types like `Map`, `function`, `Set`, etc.

Reviewer: ```suggestion
- The Vercel Edge middleware will always be the **first** function to receive the `Request` and the last function to receive `Response`. This an architectural constraint that follows the [boundaries set by Vercel](https://vercel.com/docs/concepts/functions/edge-middleware).
- Only `request` and `context` may be used to produce an `Astro.locals` object. Operations like redirects, etc. should be delegated to Astro middleware.
- `Astro.locals` **must be serializable**. Failing to do so will result in a **runtime error**. This means that you **cannot** store complex types like `Map`, `function`, `Set`, etc.
```
I worry about what this will look like visually, so suggest making these all in the bullet points for now. If it needs changing, we can do it after we see it in docs.
